clients/client/src/main/java/org/projectnessie/client/api/PagingBuilder.java

```diff
@@ -15,16 +15,49 @@
  */
 package org.projectnessie.client.api;
 
-import org.projectnessie.model.EntriesResponse;
+import java.util.OptionalInt;
+import java.util.stream.Stream;
+import org.projectnessie.error.NessieNotFoundException;
 
-public interface PagingBuilder<R extends PagingBuilder<R>> {
+public interface PagingBuilder<R extends PagingBuilder<R, RESP, ENTRY>, RESP, ENTRY> {
 
-  /** Recommended: specify the maximum number of records to return. */
+  /**
+   * When using the manual paging, it is recommended to set this parameter, this parameter is
+   * configured when necessary via {@link #stream(OptionalInt)}.
+   *
+   * <p>Only for manual paging via {@link #get()} - do <em>not</em> call when using any of the
+   * {@link #stream()} functions.
+   */
   R maxRecords(int maxRecords);
 
   /**
-   * For paged requests: pass the {@link EntriesResponse#getToken()} from the previous request,
-   * otherwise don't call this method.
+   * When using the manual paging, pass the {@code token} from the previous request, otherwise don't
+   * call this method.
+   *
+   * <p>Only for manual paging via {@link #get()} - do <em>not</em> call when using any of the
+   * {@link #stream()} functions.
    */
   R pageToken(String pageToken);
+
+  /**
+   * Fetches a response chunk (might be one page or complete response depending on use case and
+   * parameters), but callers must implement paging on their own, if necessary. If in doubt, use
+   * {@link #stream()} instead.
+   */
+  RESP get() throws NessieNotFoundException;
+
+  /**
+   * This is a convenience function that is equivalent to {@link #stream(OptionalInt)} with default
+   * page-size-hint and no limit on the number of returned entries.
+   */
+  default Stream<ENTRY> stream() throws NessieNotFoundException {
+    return stream(OptionalInt.empty());
+  }
+
+  /**
+   * Retrieve entries/results as a Java {@link Stream}, uses automatic paging.
+   *
+   * @param maxTotalRecords {@link OptionalInt#empty()} for unlimited amount of results or a limit
+   */
+  Stream<ENTRY> stream(OptionalInt maxTotalRecords) throws NessieNotFoundException;
```

Comment:
Maybe rename `maxTotalRecords` to `limit` so as to avoid confusion with `maxRecords()`?

On the other hard, we enforce stream limits only on the client side, so there's no real benefit in making it an API parameter ATM. Do you envision us passing it down to the server? Pehaps we could leave it out for now, keep only the parameterless `stream()` variant and add limits later, if/when they can be used to improve I/O?

For now a user can always do `builder.stream().limit(myLimit)`... WDYT?